fuse/src/main/java/tachyon/fuse/TachyonFuseFs.java

```diff
@@ -0,0 +1,577 @@
+
+import tachyon.Constants;
+import tachyon.TachyonURI;
+import tachyon.client.file.TachyonFile;
+import tachyon.client.file.TachyonFileSystem;
+import tachyon.exception.FileAlreadyExistsException;
+import tachyon.exception.FileDoesNotExistException;
+import tachyon.exception.InvalidPathException;
+import tachyon.exception.TachyonException;
+import tachyon.thrift.FileInfo;
+
+import jnr.ffi.Pointer;
+import jnr.ffi.types.mode_t;
+import jnr.ffi.types.off_t;
+import jnr.ffi.types.size_t;
+import ru.serce.jnrfuse.ErrorCodes;
+import ru.serce.jnrfuse.FuseFillDir;
+import ru.serce.jnrfuse.FuseStubFS;
+import ru.serce.jnrfuse.struct.FileStat;
+import ru.serce.jnrfuse.struct.FuseFileInfo;
+
+import static jnr.constants.platform.OpenFlags.O_RDONLY;
+import static jnr.constants.platform.OpenFlags.O_WRONLY;
+
+/**
+ * Main FUSE implementation class.
+ *
+ * Implements the FUSE callbacks defined by jnr-fuse
+ */
+public final class TachyonFuseFs extends FuseStubFS {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  private static final int MAX_OPEN_FILES = Integer.MAX_VALUE;
+  // Limits the number of translated (FUSE to TachyonURI)  paths that are kept
+  // in memory
+  private static final int MAX_CACHED_PATHS = 500;
+  private static final long[] UID_AND_GID = TachyonFuseUtils.getUidAndGid();
+
+  private final TachyonFileSystem mTFS;
+  private final Path mMountPoint;
+  private final Path mTachyonRootPath;
+  private final String mTachyonMaster;
+  // Keeps a cache of the most recently translated paths from String to TachyonURI
+  private final LoadingCache<String, TachyonURI> mPathResolverCache;
+
+  private final Object mOpenFilesLock;
+  // Table of open files with corresponding InputStreams and OutptuStreams
+  private final Map<Long, OpenFileEntry> mOpenFiles;
+  private long mOpenFileIds;
+
+  public TachyonFuseFs(TachyonFuseOptions opts) {
+    super();
+    mTFS = TachyonFileSystem.TachyonFileSystemFactory.get();
+    mMountPoint = Paths.get(opts.getMountPoint());
+    mTachyonMaster = opts.getMasterAddress();
+    mTachyonRootPath = Paths.get(opts.getTachyonRoot());
+    mOpenFileIds = 0L;
+    mOpenFiles = new HashMap<Long, OpenFileEntry>();
+    mOpenFilesLock = new Object();
+
+    mPathResolverCache = CacheBuilder.newBuilder()
+        .maximumSize(MAX_CACHED_PATHS)
+        .build(new PathCacheLoader());
+
+    Preconditions.checkArgument(mMountPoint.isAbsolute(),
+        "mount point should be an absolute path");
+    Preconditions.checkArgument(mTachyonRootPath.isAbsolute(),
+        "tachyon root path should be absolute");
+  }
+
+  @Override
+  public int create(String path, @mode_t long mode, FuseFileInfo fi) {
+    int ret = 0;
+    final TachyonURI turi = mPathResolverCache.getUnchecked(path);
+    final int flags = fi.flags.get();
+    LOG.trace("create({}, {}) [Tachyon: {}]", path, Integer.toHexString(flags), turi);
+    LOG.warn("{}: mode is ignored in tachyon-fuse", path);
```

Comment:
It's unfortunate that creating a file will always generate this warning since there's nothing the user can do about it. Do you think it makes more sense to put this warning in the docs instead of here?